### PR TITLE
`critical_section` implementations & esp_backtrace

### LINKED ISSUE #87: `critical_section` implementations
With the upcoming release of `critical-section`, default implementations for each arch are being removed. Instead, they should be implemented by the HALs/PAC on a per chip basis. This allows critical sections to be correctly implemented for dual-core chips, and or take advantage of hardware semaphores instead of spin locking.

A note on Xtensa implementations, we should __not__ be disabling interrupts for our critical sections, this means the `waiti` instruction will never return. Instead, we should be raising the `PS.INTLEVEL` accordingly. See https://github.com/esp-rs/xtensa-lx/issues/20.

The new release isn't out yet, but we should track the implementations we'll need here.

- [ ] Multicore aware Xtensa cs - esp32, esp32s3
- [ ] RISCV32 single core cs - esp32c3
- [ ] Xtensa single core cs - esp32s2

### PR DESCRIPTION
**Pull request by MabezDev**
Closes #87 

TODO:

- [x] Rework examples for esp32
- [x] Rework examples for esp32s3
- [x] Remove unnecessary unsafe & static mut
- [x] Test works in esp-wifi

### CODE REVIEW

File: esp-hal-common/src/lib.rs [hunk 1]
```diff
@@ -101,3 +102,118 @@ pub fn get_core() -> Cpu {
...
+
+            unsafe fn release(token: critical_section::RawRestoreState) {
+                if token != 0 {
+                    #[cfg(feature = "multicore")]
+                    {
+                        debug_assert!(multicore::MULTICORE_LOCK.is_owned_by_current_thread());
+                        // safety: we logically own the mutex from acquire()
+                        multicore::MULTICORE_LOCK.force_unlock();
+                    }
+                    core::arch::asm!("wsr.ps {0}", in(reg) token)
```
Comment (gustavonihei) [inline esp-hal-common/src/lib.rs:134]: I believe it is important to add an `rsync` here to ensure the value is committed to the PS special register.

**Comment by bjoernQ** [comment]
Tested with `esp-wifi`and got it to work. Looking forward to retire `SpinLockMutex` and `CriticalSectionMutex` 😄

**Comment by MabezDev** [comment]
This is now ready for review! Takes care of the TODO items above, but I also took the liberty of switching to esp-backtrace as I was touching every example anyways.

The remaining CI error seems spurious, no idea why it can't find the backtrace crate. Works fine locally.

**Review by bjoernQ — APPROVED** [review]
LGTM

**Comment by bjoernQ** [comment]
Strange thing about cargo being unable to find esp-backtrace - I tried to re-run the job twice but still the same 🤷‍♂️ 
@jessebraham any idea?

**Comment by jessebraham** [comment]
Sorry, I didn't mean to push to your PR, had the wrong upstream set 😅 

I was going to propose this commit be added, but I've bumped the MSRV to 1.60.0 and it seems to have solved the issue locally at least. Hopefully the CI succeeds. @bjoernQ any complaints with this? Still gives us 3 stable versions.

**Review by jessebraham — APPROVED** [review]

**Pull request merged**